src/hangar/external/manage_plugins.py

```diff
@@ -0,0 +1,219 @@
+from collections import defaultdict
+import configparser
+import os
+from glob import glob
+from warnings import warn
+from typing import Callable, Tuple
+
+
+# TODO: test cases
+
+
+def _clear_plugins():
+    """
+    Clear the plugin state to the default, i.e., where no plugins are loaded
+    """
+    plugin_details_store.clear()
+    plugin_provides.clear()
+    plugin_module_name.clear()
+    plugin_meta_data.clear()
+    _plugin_store.clear()
+
+
+def _scan_plugins():
+    """
+    Scan the plugins directory for .ini files and parse them
+    to autofill storage
+    """
+    pd = os.path.dirname(__file__)
+    config_files = glob(os.path.join(pd, '_plugins', '*.ini'))
+
```

Comment:
So how does someone add a plugin? The files need to go into the hangar/external/_plugins directory?

Reply:
I'm only skimming comments, but if plugins don't support something like "install this plugin from pypi; activate by calling `somepluginpkg.activate()`" then the plugin system is DOA for use cases like mine (which require that we get wheels and check the SHA before we use them in our build system).

Reply:
OK, I am sorry I did not touch this part. This was the existing logic and we probably need to move it to user home directory or something. And yes @elistevens, the idea is to copy the required files on `pip install` or something like `plugin.activate`. The plan is to make a cookie-cutter package/repo that does this and plugin developers need to work only on the methods such as `load`
Now that when I think about it, the whole copy thing could be avoided by having each plugin provides some convenient way to let hangar knows what functions are they exposing and what file format they could work with

Reply:
Sorry, I was unclear: my usage absolutely will not allow files in the virtual env to be modified at runtime.

Have a function to register a plugin that takes a class with some callback methods or something like that.

Reply:
Okey, We can do without the copy. I did not do much on the plugin registration and discovery part and was trying to make it actually pluggable for all data types (earlier the plugin system was explicitly coded for image data).

I am considering using package metadata for plugin discovery now which avoids the need for the copy and will be registered on `pip install` itself
https://packaging.python.org/guides/creating-and-discovering-plugins/#using-package-metadata. So the flow is going to be like this

- cookie cutter repo will have file structure and setup.py which does what is required for plugin discovery
- Developers who create plugin should create a class and should inherit from `hangar.external.BasePlugin`. This parent class will have the required functions for letting the hangar knows what functions are exposed through the plugin and what are the file formats it could work with
- A plugin can expose `load`, `save`, `show` and `board_show`. Where `load` is for loading the data from the disk and return numpy array, `save` takes a numpy array from hangar repo and save that to the provided path, `show` takes numpy array from hangar repo and show like what matplotlib show does, `board_show` takes numpy array from hangar repo and returns in a form which is understood by javascript to display (this is for the hangar board)

Am I clear enough?

Reply:
So me and sherin have discussed this, and we are going to use `entry_points` and a system similar to https://github.com/mitsuhiko/pluginbase to enable and activate extensions. This should solve the issues you mentioned @elistevens. Thanks for the feedback!